hooks/cascading-scans/hook/reverse-matches.ts

```diff
@@ -0,0 +1,201 @@
+    }
+
+    return operatorFunction(props);
+  }
+
+  function templateValue(value: string): [string, boolean] {
+    if (value === undefined) return [undefined, true];
+    let mapped = Mustache.render(value, {
+      $: {
+        ...scopeLimiterAliases
+      }
+    });
+    if (mapped == "") {
+      mapped = value;
+    }
+    let rendered = Mustache.render(mapped, finding);
+    return [rendered, rendered != ""];
+  }
+
+  return [
+    scopeLimiter.allOf !== undefined ? scopeLimiter.allOf.every(validateRequirement) : true,
+    scopeLimiter.anyOf !== undefined ? scopeLimiter.anyOf.some(validateRequirement) : true,
+    scopeLimiter.noneOf !== undefined ? !scopeLimiter.noneOf.some(validateRequirement) : true,
+  ].every(entry => entry === true);
+}
+
+interface Operands {
+  lhs: string,
+  rhs: string[],
+}
+
+interface OperatorFunctions {
+  operator: (operands: Operands) => boolean,
+  validator: (operands: Operands) => void,
+}
+
+const defaultValidator: OperatorFunctions["validator"] = props => validate(props, false, false);
+
+const operatorFunctions: { [key in ScopeLimiterRequirementOperator]: OperatorFunctions } = {
+  [ScopeLimiterRequirementOperator.In]: {
+    operator: operatorIn,
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.NotIn]: {
+    operator: props => !operatorIn(props),
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.Exists]: {
+    operator: operatorExists,
+    validator: props => validate(props, true, true),
+  },
+  [ScopeLimiterRequirementOperator.DoesNotExist]: {
+    operator: props => !operatorExists(props),
+    validator: props => validate(props, true, true),
+  },
+  [ScopeLimiterRequirementOperator.Contains]: {
+    operator: operatorContains,
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.DoesNotContain]: {
+    operator: props => !operatorContains(props),
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.InCIDR]: {
+    operator: operatorInCIDR,
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.NotInCIDR]: {
+    operator: props => !operatorInCIDR(props),
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.SubdomainOf]: {
+    operator: operatorSubdomainOf,
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.NotSubdomainOf]: {
+    operator: props => !operatorSubdomainOf(props),
+    validator: defaultValidator,
+  },
+}
+
+function validate({lhs, rhs}: Operands, lhsUndefinedAllowed, rhsUndefinedAllowed) {
+  if (!lhsUndefinedAllowed && lhs === undefined) {
+    throw new Error(`annotation may not be undefined`)
+  }
+  if (!rhsUndefinedAllowed && rhs === undefined) {
+    throw new Error(`values may not be undefined`)
+  }
+}
+
+function operatorIn({lhs, rhs}: Operands): boolean {
+  return rhs.includes(lhs);
+}
+function operatorExists({lhs, rhs}: Operands): boolean {
+  return lhs !== undefined;
+}
```

Comment:
This is coming back to the point of "calling the variables `lhs` / `rhs` makes this really hard to understand", but the way I understand things, `rhs` is the list of findings that are templated based on the "values" field in the scope limiter, while `lhs` is the value from the referenced limiter annotation. So, in which case can the value from the annotation be undefined? Or rather, in which case would it be useful to test for it, since this value is fixed? Should this be `rhs !== undefined`?

Reply:
It is indeed testing whether the annotation is undefined. In other words, whether the specified annotation exists on the scan.

With some further thoughts, I can't think of a scenario where you would want to use this. You define them both before you start the scan..

I made this because the operator also exists in default Kubernetes label selector. We can delete it if you can't think of a good use-case.

Reply:
Yeah, I don't really know any good use case for that as long as adding additional scope limiters during the cascade is not possible. We should probably remove it, here and in the documentation PR.

Reply:
3f1dbd5f - Get rid of redundant operators